### ISSUE TITLE
UnnecessaryInnerClass false positive when inner class references another inner class 

### ISSUE THREAD

**Issue by matejdro** [detekt/detekt#7598]
## Expected Behavior

`UnnecessaryInnerClass` should not be raised

## Observed Behavior

`UnnecessaryInnerClass` is raised 

## Steps to Reproduce
Perform a detekt check on the following code:

```
class A(val input: Int) {
   inner class B() {
      val c = C()
   }

   inner class C() {
      val processedInput = input + 1
   }
}
```

detekt will report that `inner` on `B` is unnecessary.

## Context

Removing `inner` from the class `B` in above example will cause code to not compile, since class `B` would no longer have access to class `C`

## Your Environment

* Version of detekt used: 1.23.6
* Version of Gradle used (if applicable): 8.9
* Operating System and version: Linux 6.10.6-arch1-1

**Comment by Sanu5** [comment]
we can supress the UnnecessaryInnerClass
@Suppress("UnnecessaryInnerClass")
    inner class ReportsUnnecessaryInnerClassRule

**Comment by atulgpt** [comment]
I think more or less this TCs should pass
```kotlin
@Nested
inner class `inner reffering other inner class` {
    @Test
    fun `reports with nested empty classes`() {
        val code = """
        class A {
            inner class B {
                inner class C
            }
        }
    """.trimIndent()

        val findings = subject.compileAndLintWithContext(env, code)
        assertThat(findings)
            .hasSize(2)
            .map(Finding::message)
            .contains(
                tuple("Class 'C' does not require `inner` keyword."),
                tuple("Class 'B' does not require `inner` keyword."),
            )
    }

    @Test
    fun `reports when only outer ctor is used`() {
        val code = """
        class A {
            inner class B {
                val a = A()
            }
        }
    """.trimIndent()

        val findings = subject.compileAndLintWithContext(env, code)
        assertThat(findings)
            .hasSize(1)
    }

    @Test
    fun `does not report when inner class refers other inner class which is required`() {
        val code = """
            class A {
                val foo = "BAR"
                inner class B {
                    val c = C()
                }

                inner class C(val fizz: String = foo)
            }
    """.trimIndent()

        assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
    }

    @Test
    fun `does not report when inner class refers other inner class in init block`() {
        val code = """
            class A {
                val foo = "BAR"
                inner class B {
                    init {
                        val c = C()
                    }
                }

                inner class C(val fizz: String = foo)
            }
        """.trimIndent()

        assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
    }

    @Test
    fun `does not report when inner class which refers other inner class C which refers inner class D which is required`() {
        val code = """
            class A {
                val foo = "BAR"
                inner class B {
                    val c = C()
                }

                inner class C {
                    val d = D()
                }

                inner class D(val fizz: String = foo)
            }
        """.trimIndent()

        assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
    }

    @Test
    fun `does not report when inner class refers empty inner class`() {
        val code = """
            class A {
                val foo = "BAR"
                inner class B {
                    val c = C()
                }

                inner class C
            }
        """.trimIndent()

        assertThat(subject.compileAndLintWithContext(env, code))
            .hasSize(1)
            .first()
            .hasMessage("Class 'C' does not require `inner` keyword.")
    }

    @Test
    fun `does report when nested usage of inner class for not required grand parent inner class`() {
        val code = """
            class A {
                // this inner is not required as C and D can be internally access the each other
                // constructors
                inner class B {
                    inner class C {
                        val d = D()
                    }

                    inner class D {
                        val c = C()
                    }
                }
            }
        """.trimIndent()

        assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
    }
}
```

**Comment by atulgpt** [comment]
@BraisGabin this can be closed as after migration to AA these cases are not failing I have added a PR https://github.com/detekt/detekt/pull/8504 to add those cases
